### PR TITLE
Implement AbstractPlugin API

### PR DESCRIPTION
**Pull request by deathaxe**
superseeds #5 

This PR implements the `LemminxPlugin` based on the `AbstractPlugin` API.

The _server.json_ is replaced by hardcoded URL and SHA256 checksum to reduce efforts to install the server binary.

All entry points to edit the _LSP-lemminx.sublime-settings_ are removed as it intended to be a provider-only plugin.

It uses the new `DottedDict` settings structure.

### CODE REVIEW

File: plugin.py [hunk 1]
```diff
@@ -1,180 +1,80 @@
...
 import sublime
 
-from sublime_lib import ActivityIndicator
+from urllib.request import urlretrieve
 
-from LSP.plugin.core.handlers import LanguageHandler
-from LSP.plugin.core.settings import ClientConfig, read_client_config
+from LSP.plugin.core.sessions import AbstractPlugin
+from LSP.plugin.core.sessions import register_plugin
+from LSP.plugin.core.sessions import unregister_plugin
```
Comment (rwols) [inline plugin.py]: You should use `from LSP.plugin import ...`, I'll try to keep `plugin/__init__.py` stable (or deprecate things in advance in that file)

**Comment by rchl** [comment]
Why do you think that it's not worth to make it easy for the user to edit (customize) settings? It seems like this server has some that user might want to change.

**Comment by rchl** [comment]
Is the plan to have a separate release for ST4? Otherwise it will break compatibility with ST3 (obviously).

File: plugin.py [hunk 2]
```diff
@@ -1,180 +1,80 @@
...
+        return "lemminx"
 
-        else:
-            yield key, value
+    @classmethod
+    def configuration(cls):
+        name = "LSP-" + cls.name()
+        base_name = name + ".sublime-settings"
+        file_path = "Packages/{}/{}".format(__package__, base_name)
+        settings = sublime.load_settings(base_name)
```
Comment (rchl) [inline plugin.py:28]: You can just call `super().configuration()` for that part.

File: plugin.py [hunk 3]
```diff
@@ -1,180 +1,80 @@
...
-        if not cls.ready:
-            cls.thread = threading.Thread(target=cls.download)
-            cls.thread.start()
-
+    def install_or_update(cls):
         # clear old server binaries
-        for fn in os.listdir(dest_path):
-            fp = os.path.join(dest_path, fn)
+        for file_name in os.listdir(cls.server_dir):
+            if file_name.splitext()[1].lower() != ".jar":
```
Comment (rchl) [inline plugin.py]:
```
Unable to start lemminx
Traceback (most recent call last):
  File "/Users/me/Library/Application Support/Sublime Text/Packages/LSP/plugin/core/windows.py", line 215, in start_async
    plugin_class.install_or_update()
  File "/Users/me/Library/Application Support/Sublime Text/Packages/LSP-lemminx/plugin.py", line 58, in install_or_update
    if file_name.splitext()[1].lower() != ".jar":
AttributeError: 'str' object has no attribute 'splitext'
```

File: plugin.py (hunk 2 above)
Comment (deathaxe) [reply plugin.py:28]: It's not the same.

File: plugin.py (hunk 3 above)
Comment (deathaxe) [reply plugin.py]: Oups. Unfortunatelly never came to that point to see that failing as ST freezes before.

**Pull request closed without merging**